### PR TITLE
[AIRFLOW-1899] Fix Kubernetes tests

### PR DESCRIPTION
**Pull request by Fokko**
Make sure you have checked _all_ steps below.

@dimberman 

What do you think? I've missed your integration tests, but I was doing the same thing in parallel. I've created some tests that use the json-rest api to start a job, and then assert the state. Currently there is one sunny-day scenario, which is ok for now. 

cc @bolkedebruin I also had some issues with timezones. Postgres complained that I needed to provide a timezone.

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-1899\] My Airflow PR"
    - https://issues.apache.org/jira/browse/AIRFLOW-1899
    - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a JIRA issue.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"

### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

File: airflow/contrib/kubernetes/worker_configuration.py [hunk 1]
```diff
@@ -21,13 +21,18 @@
 
 from airflow.contrib.kubernetes.pod import Pod, Resources
 from airflow.contrib.kubernetes.secret import Secret
+from airflow.utils.log import logging_mixin
```
Comment (bolkedebruin) [inline airflow/contrib/kubernetes/worker_configuration.py]: why not `LoggingMixin` directly?

File: airflow/contrib/kubernetes/worker_configuration.py [hunk 2]
```diff
@@ -76,10 +81,10 @@ def _get_environment(self):
         """Defines any necessary environment variables for the pod executor"""
         env = {
             'AIRFLOW__CORE__DAGS_FOLDER': '/tmp/dags',
-            'AIRFLOW__CORE__EXECUTOR': 'LocalExecutor'
+            'AIRFLOW__CORE__EXECUTOR': 'KubernetesExecutor'
```
Comment (bolkedebruin) [inline airflow/contrib/kubernetes/worker_configuration.py]: dbl check? Is this correct?

File: airflow/contrib/kubernetes/worker_configuration.py [hunk 3]
```diff
@@ -172,6 +177,10 @@ def make_pod(self, namespace, worker_uuid, pod_id, dag_id, task_id, execution_da
         annotations = {
             'iam.cloud.google.com/service-account': gcp_sa_key
         } if gcp_sa_key else {}
+        airflow_command = airflow_command.replace("-sd", "-i -sd")
```
Comment (bolkedebruin) [inline airflow/contrib/kubernetes/worker_configuration.py:180]: Please state your intention here. Why arew you manipulating the cmdline?

File: airflow/models.py [hunk 4]
```diff
@@ -1110,6 +1110,15 @@ def current_state(self, session=None):
             state = None
         return state
 
+    @provide_session
+    def delete(self, session=None):
```
Comment (bolkedebruin) [inline airflow/models.py]: Is this related?

File: airflow/www_rbac/api/experimental/endpoints.py [hunk 5]
```diff
@@ -64,7 +66,7 @@ def trigger_dag(dag_id):
         except ValueError:
             error_message = (
                 'Given execution date, {}, could not be identified '
-                'as a date. Example date format: 2015-11-16T14:34:15+00:00'
+                'as a date. Example date format: 2015-11-16T14:34:15Z'
```
Comment (bolkedebruin) [inline airflow/www_rbac/api/experimental/endpoints.py]: Both were correct. +00:00 equals Z. Why this change?

File: airflow/www_rbac/views.py [hunk 6]
```diff
@@ -1113,7 +1113,7 @@ def duration(self, session=None):
             base_date = dag.latest_execution_date or timezone.utcnow()
 
         dates = dag.date_range(base_date, num=-abs(num_runs))
-        min_date = dates[0] if dates else datetime(2000, 1, 1)
+        min_date = dates[0] if dates else datetime(2000, 1, 1, tzinfo=timezone.utc)
```
Comment (bolkedebruin) [inline airflow/www_rbac/views.py]: don't use datetime directly and dont set to UTC hardcoded. Use "timezone.datetime" instead of "datetime.datetime"

File: airflow/www_rbac/views.py [hunk 7]
```diff
@@ -1216,7 +1216,7 @@ def tries(self, session=None):
...
```
Comment (bolkedebruin) [inline airflow/www_rbac/views.py]: Idem

File: airflow/www_rbac/views.py [hunk 8]
```diff
@@ -1279,7 +1279,7 @@ def landing_times(self, session=None):
...
```
Comment (bolkedebruin) [inline airflow/www_rbac/views.py]: Idem

File: run_unit_tests.sh [hunk 9]
```diff
@@ -44,6 +44,7 @@ nose_args=$@
 which airflow > /dev/null || python setup.py develop
 
 echo "Initializing the DB"
+airflow initdb
```
Comment (bolkedebruin) [inline run_unit_tests.sh]: Isn;t this redundant with below?

File: tests/cli/test_cli.py [hunk 10]
```diff
@@ -74,3 +142,23 @@ def test_cli_webserver_debug(self):
...
         p.terminate()
         p.wait()
+
+    def test_local_run(self):
+        args = create_mock_args(
+            task_id='print_the_context',
+            dag_id='example_python_operator',
+            subdir='/root/dags/example_python_operator.py',
+            interactive=True,
+            execution_date=Pendulum.parse('2018-04-27T08:39:51.298439+00:00')
```
Comment (bolkedebruin) [inline tests/cli/test_cli.py]: for tests this is ok, but you can also use `timezone.parse`

**Review by bolkedebruin — CHANGES_REQUESTED** [review]
I would split the non related (but important fixes) changes to Kubernetes away. Some smaller nits, but please fix the datetime usage to use the correct API.

File: airflow/contrib/kubernetes/worker_configuration.py (hunk 1 above)
Comment (Fokko) [reply airflow/contrib/kubernetes/worker_configuration.py]: Thanks

File: airflow/contrib/executors/kubernetes_executor.py [hunk 11]
```diff
@@ -114,40 +104,37 @@ def __init__(self):
...
         # The branch of the repository to be checked out
-        self.git_branch = self.safe_get(self.kubernetes_section, 'git_branch', None)
+        self.git_branch = conf.get(self.kubernetes_section, 'git_branch')
         # Optionally, the directory in the git repository containing the dags
-        self.git_subpath = self.safe_get(self.kubernetes_section, 'git_subpath', '')
+        self.git_subpath = conf.get(self.kubernetes_section, 'git_subpath')
 
         # Optionally a user may supply a `git_user` and `git_password` for private
         # repositories
-        self.git_user = self.safe_get(self.kubernetes_section, 'git_user', None)
```
Comment (dimberman) [inline airflow/contrib/executors/kubernetes_executor.py:135]: So it's worth mentioning that we used to have our own "safe_get" in the kube_config because of configs like this one. Git mode is one of two options and we didn't want the system to fail if it wasn't there. Would users just have it empty in their .cfg if they're not using it?

File: airflow/models.py (hunk 4 above)
Comment (Fokko) [reply airflow/models.py]: No, I think this is a helper that @dimberman created

File: airflow/www_rbac/api/experimental/endpoints.py (hunk 5 above)
Comment (Fokko) [reply airflow/www_rbac/api/experimental/endpoints.py]: This was a WIP. I got some timezone exceptions, but in the end it was unrelated from this format.

File: airflow/contrib/kubernetes/worker_configuration.py (hunk 2 above)
Comment (dimberman) [reply airflow/contrib/kubernetes/worker_configuration.py]: Technically the executor shouldn't matter if we're running the task in local mode. I was changing around the executor to see if we could figure out the source of some of the logging/local task bugs.

File: airflow/www_rbac/api/experimental/endpoints.py (hunk 5 above)
Comment (Fokko) [reply airflow/www_rbac/api/experimental/endpoints.py]: https://github.com/apache/incubator-airflow/pull/3303

File: airflow/contrib/kubernetes/worker_configuration.py (hunk 3 above)
Comment (dimberman) [reply airflow/contrib/kubernetes/worker_configuration.py:180]: I was trying to start the task in interactive mode (should've been -int rather than -i) to see if that would fix the bug where logs seem to die and not offer any feedback to the running of the task.

File: airflow/models.py (hunk 4 above)
Comment (dimberman) [reply airflow/models.py]: No this can be deleted. I was only using it for testing.

File: run_unit_tests.sh (hunk 9 above)
Comment (dimberman) [reply run_unit_tests.sh]: I was running into some strangeness where it would attempt to reset a db that had not yet been initialized.

File: scripts/ci/kubernetes/kube/airflow.yaml [hunk 12]
```diff
@@ -61,7 +61,7 @@ spec:
           - "bash"
         args:
           - "-cx"
-          - "cd /usr/local/lib/python2.7/dist-packages/airflow && cp -R example_dags/* /root/airflow/dags/ && airflow initdb && alembic upgrade heads"
+          - "./tmp/airflow-init.sh"
```
Comment (dimberman) [inline scripts/ci/kubernetes/kube/airflow.yaml:64]: +1 for this. definitely much cleaner.

File: tests/contrib/minikube/test_kubernetes_executor.py [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+    raise unittest.SkipTest(
+        "Kubernetes integration tests require a minikube cluster;"
+        "Skipping tests {}".format(e)
+    )
+
+
+class KubernetesExecutorTest(unittest.TestCase):
+
+    def test_integration_run_dag(self):
+        host_ip = check_output(['minikube', 'ip'])
```
Comment (dimberman) [inline tests/contrib/minikube/test_kubernetes_executor.py:38]: maybe this can be in a before_all step so it won't need to be repeated for future tests?

File: tests/contrib/minikube/test_kubernetes_executor.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+        self.assertGreater(len(result_json['items']), 0)
+
+        execution_date = result_json['items'][0]['execution_date']
+        print("Found the job with execution date {}".format(execution_date))
+
+        tries = 0
+        state = ''
+        # Wait 100 seconds for the operator to complete
+        while tries < 20:
```
Comment (dimberman) [inline tests/contrib/minikube/test_kubernetes_executor.py:74]: I think this could be separated into a "dag_succeeded" function for re-use.

File: tests/contrib/minikube_tests/integration/test_kubernetes_executor_integration.py [hunk 15]
```diff
@@ -1,67 +0,0 @@
...
-                     'Kubernetes integration tests are unsupported by this configuration')
-    def test_kubernetes_executor_dag_runs_successfully(self):
-        dag_id, run_id = "example_python_operator", uuid4().hex
-        run_dag(dag_id, run_id)
-        state = dag_final_state(dag_id, run_id, timeout=120)
-        self.assertEquals(state, DagRunState.SUCCESS)
-
-    @unittest.skipIf(SKIP_KUBE,
-                     'Kubernetes integration tests are unsupported by this configuration')
-    def test_start_dag_then_kill_scheduler_then_ensure_dag_succeeds(self):
```
Comment (dimberman) [inline tests/contrib/minikube_tests/integration/test_kubernetes_executor_integration.py:42]: I think that testing the Scheduler killing + restarting is a pretty important feature for the executor. Could you please re-implement this test in the integration tests?

**Review by dimberman — CHANGES_REQUESTED** [review]

File: airflow/www_rbac/views.py (hunk 6 above)
Comment (Fokko) [reply airflow/www_rbac/views.py]: Separate PR in https://github.com/apache/incubator-airflow/pull/3303

File: airflow/www_rbac/views.py (hunk 7 above)
Comment (Fokko) [reply airflow/www_rbac/views.py]: Separate PR in https://github.com/apache/incubator-airflow/pull/3303

File: airflow/www_rbac/views.py (hunk 8 above)
Comment (Fokko) [reply airflow/www_rbac/views.py]: Separate PR in https://github.com/apache/incubator-airflow/pull/3303

**Comment by dimberman** [comment]
I think it's a really good idea to use the airflow API rather than kubernetes commands. I definitely still want the "kill scheduler then resume" test to remain since resuming tasks is really important.

File: tests/cli/test_cli.py (hunk 10 above)
Comment (Fokko) [reply tests/cli/test_cli.py]: Replaced it with `timezone.parse`

**Comment by Fokko** [comment]
@dimberman We're still using the API: https://github.com/Fokko/incubator-airflow/blob/9c5ce551013cb46b58584e51e04ac358a2e48b27/tests/contrib/minikube/test_kubernetes_executor.py

I agree with you that we should test this indeed.

**Comment by Fokko** [comment]
Thinking about it @dimberman the test that you mention shouldn't be too hard to implement and rather valuable.

**Comment by Fokko** [comment]
@bolkedebruin Any further comments? I've split some unrelated changes into separate PR's and polished it a bit.

**Comment by dimberman** [comment]
@Fokko sounds good. Could you please @ me in the other PRs? I've been fighting off a cold the last few days but should have time tomorrow to get some work done (And now that I know where the logs are I suspect it'll be a lot easier to work).

**Comment by Fokko** [comment]
Yes, of course! Take care!
